File: Submariner/SBSubsonicDownloadOperation.m
//
//  SBSubsonicDownloadOperation.m
//  Submariner
//
//  Created by Rafaël Warnault on 16/06/11.
//
//  Copyright (c) 2011-2014, Rafaël Warnault
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of the Read-Write.fr nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "SBSubsonicDownloadOperation.h"
#import "SBImportOperation.h"

#import "SBLibrary.h"
#import "SBTrack.h"
#import "SBServer.h"
#import "SBOperationActivity.h"

#import "NSURL+Parameters.h"
#import "NSOperationQueue+Shared.h"
#import "NSManagedObjectContext+Fetch.h"
#import "NSString+Hex.h"
#import "NSString+File.h"




NSString *SBSubsonicDownloadStarted     = @"SBSubsonicDownloadStarted";
NSString *SBSubsonicDownloadFinished    = @"SBSubsonicDownloadFinished";





@interface SBSubsonicDownloadOperation (Private)
- (void)startDownloadingURL:(NSURL *)url;
@end





@implementation SBSubsonicDownloadOperation


@synthesize trackID;
@synthesize activity;


- (id)initWithManagedObjectContext:(NSManagedObjectContext *)mainContext
{
    self = [super initWithManagedObjectContext:mainContext];
    if (self) {
        // Initialization code here.
        SBLibrary *library = (SBLibrary *)[[self mainContext] fetchEntityNammed:@"Library" withPredicate:nil error:nil];
        libraryID = [library objectID];
        
        activity = [[SBOperationActivity alloc] init];
        [self.activity setIndeterminated:YES];
        
        [[NSNotificationCenter defaultCenter] postNotificationName:SBSubsonicDownloadStarted
                                                            object:self.activity];
    }
    
    return self;
}




- (void)finish {
    [[NSNotificationCenter defaultCenter] postNotificationName:SBSubsonicDownloadFinished
                                                        object:self.activity];
    
    [super finish];
}


- (void)main {
    @autoreleasepool {
    
        SBTrack *track = (SBTrack *)[[self threadedContext] objectWithID:trackID];
        
        // prepare activity stack
        [self.activity setOperationName:[NSString stringWithFormat:@"Download « %@ »", track.itemName]];
        [self.activity setOperationInfo:@"Pending Request..."];
        
        // prepare download URL
        NSMutableDictionary *parameters = [NSMutableDictionary dictionary];
        [track.server getBaseParameters: parameters];
        [parameters setValue:track.id forKey:@"id"];
        
        // XXX: Stream URL?
        NSURL *url = [NSURL URLWithString:track.server.url command:@"rest/download.view" parameters:parameters];
        // No more issues calling this outside of main, I bet
        [self startDownloadingURL: url];
    
    }
}



- (void)startDownloadingURL:(NSURL *)url
{    
    // Create the request.
    NSURLRequest *theRequest = [NSURLRequest requestWithURL:url
                                                cachePolicy:NSURLRequestUseProtocolCachePolicy
                                            timeoutInterval:30.0];
    
    // Create the connection with the request and start loading the data.
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration delegate: self delegateQueue: nil];
    NSURLSessionDownloadTask *task = [session downloadTaskWithRequest: theRequest];

    // This was on the delegate for the old NSURLDownload begin
    [self.activity setIndeterminated:NO];
    [self.activity setOperationInfo:@"Downloading Track..."];

    [task resume];
}




#pragma mark -
#pragma mark NSURLSession delegate (Authentification)

// we don't need to implement the one without task:
-(void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential * _Nullable))completionHandler {
    // XXX: Check the authentication type if it's NSURLAuthenticationMethodHTTPBasic or NSURLAuthenticationMethodServerTrust
    if ([challenge previousFailureCount] == 0) {
        
        SBTrack *track = (SBTrack *)[[self threadedContext] objectWithID:trackID];
        
        NSURLCredential *newCredential;
        newCredential = [NSURLCredential credentialWithUser:track.server.username
                                                   password:track.server.password
                                                persistence:NSURLCredentialPersistenceNone];
        
        //[[challenge sender] useCredential:newCredential
        //       forAuthenticationChallenge:challenge];
        completionHandler(NSURLSessionAuthChallengeUseCredential, newCredential);
        
    } else {
        // XXX: Better handling here?
        //[[challenge sender] cancelAuthenticationChallenge:challenge];
        completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, nil);
    }
}


#pragma mark -
#pragma mark NSURLSession delegate (State)

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error {
    // we handle success in the didFinishDownloadingToURL callback
    if (error != nil) {
        NSLog(@"Error in NSURLSession: %@", error);
        [NSApp performSelectorOnMainThread:@selector(presentError:) withObject:error waitUntilDone:NO];
        [self finish];
        [session invalidateAndCancel];
    }
}

- (void)URLSession:(nonnull NSURLSession *)session downloadTask:(nonnull NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(nonnull NSURL *)location {
    // Do something with the data.
    [self.activity setOperationInfo:@"Importing Track..."];
    
    // We need to give it an audio extension.
    NSString *mimeType = downloadTask.response.MIMEType;
    NSString *extension = [mimeType extensionForMIMEType] ?: @"mp3"; // fallback
    NSURL *tempURL = [[NSURL temporaryFileURL] URLByAppendingPathExtension: extension];
    NSString *newTempPath = [tempURL path];
    NSError *moveError = nil;
    [[NSFileManager defaultManager] moveItemAtPath:location.path toPath: newTempPath error:&moveError];
    if (moveError) {
        NSLog(@"Error moving %@ to %@: %@", location.path, newTempPath, moveError);
    }
        
    // 3. import to library on write endx
    SBImportOperation *op = [[SBImportOperation alloc] initWithManagedObjectContext:[self mainContext]];
    [op setFilePaths:[NSArray arrayWithObject: newTempPath]];
    [op setLibraryID:libraryID];
    [op setRemoteTrackID:trackID];
    [op setCopyFile:YES];
    [op setRemove:YES];
    
    [[NSOperationQueue sharedDownloadQueue] addOperation:op];
    
    [self finish];
    [session invalidateAndCancel];
}


#pragma mark -
#pragma mark NSURLSession delegate (Progress)

- (void)URLSession:(nonnull NSURLSession *)session downloadTask:(nonnull NSURLSessionDownloadTask *)downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
    
    [self.activity setOperationCurrent:[NSNumber numberWithLongLong:totalBytesWritten]];
    [self.activity setOperationTotal:[NSNumber numberWithLongLong:totalBytesExpectedToWrite]];
    
    if (totalBytesExpectedToWrite != NSURLSessionTransferSizeUnknown) {
        
        NSString *sizeProgress = [NSString stringWithFormat:@"%.2f/%.2f MB", (float)totalBytesWritten/1024/1024, (float)totalBytesExpectedToWrite/1024/1024];
        [self.activity setOperationInfo:sizeProgress];
        // If the expected content length is
        // available, display percent complete.
        float percentComplete = (totalBytesExpectedToWrite/(float)totalBytesExpectedToWrite)*100.0;
        [self.activity setOperationPercent:[NSNumber numberWithFloat:percentComplete]];

    } else {
        // If the expected content length is
        // unknown, just log the progress.
        NSString *sizeProgress = [NSString stringWithFormat:@"%.2f MB", (float)totalBytesWritten/1024/1024];
        [self.activity setOperationInfo:sizeProgress];
        //NSLog(@"Bytes received - %ld", bytesReceived);
    }
}


@end

File: Submariner/SBSubsonicDownloadOperation.swift
//
//  SBSubsonicDownloadOperation.swift
//  Submariner
//
//  Created by Calvin Buckley on 2023-04-18.
//  Copyright © 2023 Submariner Developers. All rights reserved.
//

import Cocoa
import UniformTypeIdentifiers
import os

fileprivate let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "SBSubsonicDownloadOperation")

extension NSNotification.Name {
    static let SBSubsonicDownloadStarted = NSNotification.Name("SBSubsonicDownloadStarted")
    static let SBSubsonicDownloadFinished = NSNotification.Name("SBSubsonicDownloadFinished")
}

@objc class SBSubsonicDownloadOperation: SBOperation, URLSessionDelegate, URLSessionTaskDelegate, URLSessionDownloadDelegate {
    private let track: SBTrack
    
    let activity: SBOperationActivity
    
    @objc init!(managedObjectContext mainContext: NSManagedObjectContext!, trackID: NSManagedObjectID) {
        // Reconstitute the track because Core Data objects can't cross thread boundaries.
        track = mainContext.object(with: trackID) as! SBTrack
        
        let activityName = String.init(format: "Downloading %@%@%@",
                                       Locale.current.quotationBeginDelimiter ?? "\"",
                                       track.itemName!,
                                       Locale.current.quotationEndDelimiter ?? "\"")
        activity = SBOperationActivity(name: activityName)
        activity.operationInfo = "Pending Request..."
        activity.progress = .none
        
        super.init(managedObjectContext: mainContext)
        
        DispatchQueue.main.async {
            NotificationCenter.default.post(name: .SBSubsonicDownloadStarted, object: self.activity)
        }
    }
    
    override func main() {
        autoreleasepool {
            // We don't need to do any transformation here,
            // as downloadURL will get the auth params from SBServer.
            let url = track.downloadURL()!
            logger.info("Downloading track at URL: \(url)")
            let request = URLRequest(url: url, cachePolicy: .useProtocolCachePolicy, timeoutInterval: 30)
            let configuration = URLSessionConfiguration.default
            let session = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
            let task = session.downloadTask(with: request)
            task.resume()
        }
    }
    
    override func finish() {
        DispatchQueue.main.async {
            NotificationCenter.default.post(name: .SBSubsonicDownloadFinished, object: self.activity)
        }
        super.finish()
    }
    
    // #MARK: -
    // #MARK: NSURLSession Delegate (Auth)
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if (challenge.previousFailureCount == 0) {
            if let server = track.server, let username = server.username, let password = server.password {
                let credential = URLCredential(user: username,
                                               password: password,
                                               persistence: .none)
                
                completionHandler(.useCredential, credential)
            }
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
    
    // #MARK: -
    // #MARK: NSURLSession Delegate (State)
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        if let error = error {
            logger.error("Failure downloading track with URLSession, error \(error, privacy: .public)")
            DispatchQueue.main.async {
                NSApp.presentError(error)
            }
            self.finish()
            session.invalidateAndCancel()
        }
    }
    
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        // Success
        DispatchQueue.main.async {
            self.activity.operationInfo = "Importing track..."
        }
        
        // SBImportOperation needs an audio file extension. Rename the file.
        let fileType = UTType(mimeType: downloadTask.response?.mimeType ?? "audio/mp3") ?? UTType.mp3
        let temporaryFile = URL.temporaryFile().appendingPathExtension(for: fileType)
        try! FileManager.default.moveItem(at: location, to: temporaryFile)
        
        // Now import.
        if let importOperation = SBImportOperation(managedObjectContext: mainContext, file: temporaryFile, remoteTrackID: track.objectID) {
            OperationQueue.sharedDownloadQueue.addOperation(importOperation)
        }
        
        self.finish()
        session.finishTasksAndInvalidate()
    }
    
    // #MARK: -
    // #MARK: NSURLSession Delegate (Progress)
    
    private let byteCountFormatter = MeasurementFormatter()
    
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        let totalWritten = Measurement<UnitInformationStorage>(value: Double(totalBytesWritten), unit: .bytes).converted(to: .megabytes)
        
        if totalBytesExpectedToWrite != NSURLSessionTransferSizeUnknown {
            let totalToWrite = Measurement<UnitInformationStorage>(value: Double(totalBytesExpectedToWrite), unit: .bytes)
                .converted(to: .megabytes)
            DispatchQueue.main.async {
                self.activity.progress = .determinate(n: Float(totalBytesWritten), outOf: Float(totalBytesExpectedToWrite))
                self.activity.operationInfo = String.init(format: "Downloaded %@/%@",
                                                          self.byteCountFormatter.string(from: totalWritten),
                                                          self.byteCountFormatter.string(from: totalToWrite))
            }
        } else {
            DispatchQueue.main.async {
                self.activity.progress = .indeterminate(n: Float(totalBytesWritten))
                self.activity.operationInfo = String.init(format: "Downloaded %@",
                                                          self.byteCountFormatter.string(from: totalWritten))
            }
        }
    }
}
